### PR TITLE
feat: drawer component

### PR DESCRIPTION
**Pull request by sshanzel**
## Changes
- Drawer component.
- For the moment, this is only used on mobile so I did not introduce multiple position options such as left and right (probably useful on desktop but I don't think the design team has plans for it at the moment).
- Title property to add a pre-style header.

## Events

Did you introduce any new tracking events?
Don't forget to update the [Analytics Taxonomy sheet](https://docs.google.com/spreadsheets/d/18Lv7zXges9QfVX5VYL1a-Hyl0e1sQ3sLr0OK8YZWKXI/edit#gid=0)

Log the new/changed events below:

| Type   | event_name  | value |
|--------|-------------|-------|
| Change/New | event name  | extra: { ... } |

### On those affected packages:
- [ ] Have you done sanity checks in the webapp?
- [ ] Have you done sanity checks in the extension?
- [ ] Does this not break anything in companion?

### Did you test the modified components media queries?
- [ ] MobileL (420px)
- [ ] Tablet (656px)
- [ ] Laptop (1020px)

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

MI-124 #done

### CODE REVIEW

File: packages/shared/src/components/drawers/Drawer.tsx [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  onClose,
+}: DrawerProps): ReactElement {
+  const [hasAnimated, setHasAnimated] = useState(false);
+  const [animate] = useDebounce(() => setHasAnimated(true), 1);
+
+  useEffect(() => {
+    if (!hasAnimated) {
+      animate();
+    }
+  }, [animate, hasAnimated]);
```
Comment (kopancek) [inline packages/shared/src/components/drawers/Drawer.tsx:48]: Why do we need this code here? Seems rather hard to follow.

**Review by kopancek — COMMENTED** [review]
Looks mostly great to me. The storybook 😍 👍  

Added 1 question. Also noticed when looking at the storybook, that the element only animates when toggled on the screen, should we also have animation for toggling off - e.g. going back where it came from?

File: packages/shared/src/components/drawers/Drawer.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+interface ClassName {
+  overlay?: string;
+  drawer?: string;
+}
+
+export interface DrawerProps {
+  children: ReactNode;
+  className?: ClassName;
+  position?: DrawerPosition;
+  closeOnOutsideClick?: boolean;
```
Comment (rebelchris) [inline packages/shared/src/components/drawers/Drawer.tsx:20]: From the storybook when this is true it also closes if you click on the items, not sure this is intended behaviour?

File: packages/shared/src/components/drawers/Drawer.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+export function Drawer({
+  children,
+  className = {},
+  position = DrawerPosition.Bottom,
+  closeOnOutsideClick = true,
+  onClose,
+}: DrawerProps): ReactElement {
+  const [hasAnimated, setHasAnimated] = useState(false);
+  const [animate] = useDebounce(() => setHasAnimated(true), 1);
```
Comment (rebelchris) [inline packages/shared/src/components/drawers/Drawer.tsx:42]: Why do we need the debounce here?

File: packages/shared/src/components/drawers/Drawer.tsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    if (!hasAnimated) {
+      animate();
+    }
+  }, [animate, hasAnimated]);
+
+  return (
+    <RootPortal>
+      <div
+        className={classNames(
+          'first: fixed inset-0 z-max bg-overlay-quaternary-onion',
```
Comment (rebelchris) [inline packages/shared/src/components/drawers/Drawer.tsx:54]: `first:` does nothing here?

File: packages/shared/src/components/drawers/Drawer.tsx [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        onClick={closeOnOutsideClick && onClose}
+        role="presentation"
+      >
+        <div
+          className={classNames(
+            'absolute flex flex-col overflow-y-auto bg-theme-bg-primary px-4 py-3 transition-transform duration-200 ease-in-out',
+            !hasAnimated && animatePositionClassName[position],
+            drawerPositionToClassName[position],
+            className?.drawer,
+            'max-h-[calc(100vh-5rem)] w-full',
```
Comment (rebelchris) [inline packages/shared/src/components/drawers/Drawer.tsx:66]: What is the 5rem neg?

File: packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import React, { ReactElement, useState } from 'react';
+import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
+import { Button, ButtonVariant } from '../buttons/Button';
+import { Drawer, DrawerProps } from './Drawer';
+
+export function StorybookDrawerTrigger({
```
Comment (rebelchris) [inline packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx:6]: Can't we add this directly to storybook package?
Feels a bit misplaced now.

File: packages/storybook/stories/drawers/Drawer.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,47 @@
+import type { Meta, StoryObj } from '@storybook/react';
```
Comment (rebelchris) [inline packages/storybook/stories/drawers/Drawer.stories.tsx:1]: Are we expecting more types of Drawers?
If not I'd place it in the root to avoid the double folder structure.

**Review by rebelchris — CHANGES_REQUESTED** [review]
Wow amazing work!
Have some minor suggestions

File: packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx (hunk 6 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx:6]: It is missing quite a lot of packages, one for react-query, if we want, we can install those deps as well.

File: packages/shared/src/components/drawers/Drawer.tsx (hunk 5 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/Drawer.tsx:66]: This part. The screen should not be filled as max height.

Screenshot 2024-02-21 at 5 08 54 PM

File: packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx (hunk 6 above)
Comment (rebelchris) [reply packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx:6]: Hmm let's see what @nensidosari thinks.
Don't want to overcomplicate, might just look weird to newcomers

Comment (sshanzel) [reply packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx:6]: Yeah, agree with that.

File: packages/shared/src/components/drawers/Drawer.tsx (hunk 1 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/Drawer.tsx:48]: When animating the entrance of conditionally rendered elements, we need to ensure the element is available before running the animation itself. Otherwise, it will not work.

File: packages/shared/src/components/drawers/Drawer.tsx (hunk 3 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/Drawer.tsx:42]: Same here https://github.com/dailydotdev/apps/pull/2666#discussion_r1497149472

File: packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx (hunk 6 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx:6]: Also don't want this here. I'm open to suggestions on how we can tackle this. Maybe the right way is by installing the deps? Not sure yet at the moment 😅

File: packages/shared/src/components/drawers/Drawer.tsx (hunk 1 above)
Comment (kopancek) [reply packages/shared/src/components/drawers/Drawer.tsx]: Should we use a ref instead? How does this make sure the element is available?

File: packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx (hunk 6 above)
Comment (kopancek) [reply packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx:6]: Why do we need `QueryClientProvider` in here?

File: packages/shared/src/components/drawers/Drawer.tsx (hunk 2 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/Drawer.tsx:21]: Oh, is it. That should not be the case. Let me check quickly.

File: packages/shared/src/components/drawers/Drawer.tsx (hunk 1 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/Drawer.tsx]: On line 42, we wrapped the function in a debounce with 1ms, which ensures the elements are already written in the DOM.

We need the re-render to execute the animation, hence we can only use `useState`.

**Comment by sshanzel** [comment]
> Added 1 question. Also noticed when looking at the storybook, that the element only animates when toggled on the screen, should we also have animation for toggling off - e.g. going back where it came from?

Oh, I thought we are only animating the entrance. Yeah we can do the exit as well. Though let's check first with the design team as it will have additional engineering effort (similar to the entrance, we need to wait for the animation to finish before conditionally rendering out)

File: packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx (hunk 6 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx:6]: The Storybook is crashing without it. We have some internal hooks that utilize the query client.

File: packages/shared/src/components/drawers/Drawer.tsx (hunk 1 above)
Comment (kopancek) [reply packages/shared/src/components/drawers/Drawer.tsx]: If you use a `ref` on the main drawer element, we don't really need a debounce and it actually ensures that we have the element in the DOM. Or am I missing something? :)

Comment (sshanzel) [reply packages/shared/src/components/drawers/Drawer.tsx]: Ah, I thought you meant `useRef` over `useState`. That is actually a great idea. Let me check that quickly.

File: packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx (hunk 6 above)
Comment (kopancek) [reply packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx:6]: I think this is the reason: https://github.com/dailydotdev/apps/blob/MI-124-drawer-component/packages/shared/src/components/tooltips/Portal.tsx#L25

Do we. need to use the `RootPortal`? Maybe that could be a quick fix :)

File: packages/shared/src/components/drawers/Drawer.tsx (hunk 1 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/Drawer.tsx]: We would still need the debounce though that was a better approach since we are now one less `useEffect`. Thanks for the suggestion! 🚀

File: packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx (hunk 6 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx:6]: Arguably, should be possible not to use it. There could be some implementations in the future where the Drawer might not work without it, but we can probably tackle that when it comes.

File: packages/shared/src/components/drawers/Drawer.tsx (hunk 5 above)
Comment (rebelchris) [reply packages/shared/src/components/drawers/Drawer.tsx:65]: Oh ok got it :)

File: packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx (hunk 6 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/StorybookDrawerTrigger.tsx]: Yep. Have decided to remove the Portal and bring it back once necessary.

File: packages/shared/src/components/drawers/Drawer.tsx (hunk 2 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/Drawer.tsx:21]: Should be fixed now 🚀

Comment (rebelchris) [reply packages/shared/src/components/drawers/Drawer.tsx:21]: Yes validated! Thanks :D
Really nice to see Storybook being so easy to test now :D

**Review by rebelchris — APPROVED** [review]
Wow, love how you made it so much cleaner now and works like a charm!

File: packages/shared/src/components/drawers/Drawer.tsx (hunk 2 above)
Comment (sshanzel) [reply packages/shared/src/components/drawers/Drawer.tsx:21]: True, we are now able to say confidently that things are working 🚀

**Comment by sshanzel** [comment]
@kopancek have added the exit animation now 🥳

**Review by kopancek — APPROVED** [review]
Approving to unblock 👍

**Review by nensidosari — APPROVED** [review]
Love it!

**Comment by sshanzel** [comment]
Refactored a bit and moved away from using a hook and instead made it easier to use for everyone.
